boa/src/value/mod.rs

```diff
@@ -93,184 +128,791 @@ impl JsValue {
+}
+
+impl From<Box<str>> for JsValue {
+    #[inline]
+    fn from(string: Box<str>) -> Self {
+        From::<JsString>::from(JsString::new(string))
+    }
+}
+
+impl From<JsString> for JsValue {
+    #[inline]
+    fn from(string: JsString) -> Self {
+        let pointer = unsafe { std::mem::transmute::<JsString, *mut ()>(string) as u64 };
+        debug_assert_eq!(pointer & MASK_POINTER_PAYLOAD, pointer);
+        let value = Self {
+            value: Cell::new(STRING_TYPE | pointer),
+        };
+        debug_assert!(value.is_string());
+        debug_assert_eq!(value.tag(), ValueTag::String);
+        value
+    }
+}
+
+impl From<JsObject> for JsValue {
+    #[inline]
+    fn from(object: JsObject) -> Self {
+        let pointer = unsafe { std::mem::transmute::<JsObject, *mut ()>(object) };
+        debug_assert_eq!((pointer as u64) & MASK_POINTER_PAYLOAD, pointer as u64);
+        debug_assert_eq!(OBJECT_TYPE & MASK_POINTER_PAYLOAD, 0);
+        let value = Self {
+            value: Cell::new(OBJECT_TYPE | (pointer as u64)),
+        };
+        debug_assert!(value.is_object());
+        debug_assert_eq!(value.tag(), ValueTag::Object);
+        value
+    }
+}
+
+impl From<JsSymbol> for JsValue {
+    #[inline]
+    fn from(symbol: JsSymbol) -> Self {
+        let pointer = unsafe { std::mem::transmute::<JsSymbol, *mut ()>(symbol) as u64 };
+        debug_assert_eq!(pointer & MASK_POINTER_PAYLOAD, pointer);
+        let value = Self {
+            value: Cell::new(SYMBOL_TYPE | pointer),
+        };
+        debug_assert!(value.is_symbol());
+        debug_assert_eq!(value.tag(), ValueTag::Symbol);
+        value
+    }
+}
+
+impl From<JsBigInt> for JsValue {
+    #[inline]
+    fn from(bigint: JsBigInt) -> Self {
+        let pointer = unsafe { std::mem::transmute::<JsBigInt, *mut ()>(bigint) as u64 };
+        debug_assert_eq!(pointer & MASK_POINTER_PAYLOAD, pointer);
+        let value = Self {
+            value: Cell::new(BIGINT_TYPE | pointer),
+        };
+        debug_assert!(value.is_bigint());
+        debug_assert_eq!(value.tag(), ValueTag::BigInt);
+        value
+    }
+}
+
+impl Drop for JsValue {
+    #[inline]
+    fn drop(&mut self) {
+        unsafe {
+            match self.tag() {
+                ValueTag::Object => {
+                    let pointer = self.as_pointer();
+                    let _ = std::mem::transmute::<_, JsObject>(pointer);
+                }
+                ValueTag::String => {
+                    let pointer = self.as_pointer();
+                    let _ = std::mem::transmute::<_, JsString>(pointer);
+                }
+                ValueTag::Symbol => {
+                    let pointer = self.as_pointer();
+                    let _ = std::mem::transmute::<_, JsSymbol>(pointer);
+                }
+                ValueTag::BigInt => {
+                    let pointer = self.as_pointer();
+                    let _ = std::mem::transmute::<_, JsBigInt>(pointer);
+                }
```

Comment:
Seeing this I wonder if it would be worth to move the cast logic to each pointee type individually. I think it would be safer to use the `Rc::into_raw/from_raw` API instead of extensively using `transmute`. Also applies to the `NonNull::as_ptr/new/cast` API.
Should we abstract this as an unsafe trait, contained on a separate module?

Reply:
> Seeing this I wonder if it would be worth to delegate the cast logic to each type individually. I think it would be safer to use the `Rc::into_raw/from_raw` API instead of extensively using `transmute`. Also applies to the `NonNull::as_ptr/new/cast` API.

I thought of using `Rc::into_raw()`/`Rc::from_raw()` but it does some unnecessary computations (offsetting the pointer than giving it for encoding and offsetting it back for decoding) so i just used transmute directly (for `Gc::from_raw` its even more costly).

> Should we abstract this as an unsafe trait, contained on a separate module?

Yeah, maybe, but this was all just me experimenting, and things will change and shift around as I think off better solutions as I go. (Any advice/suggestion is welcomed as always).

Hopefully I can rebase this today  :)